File: zj_album/src/main/java/com/zj/album/ui/preview/PreviewFullValueAnimator.java
package com.zj.album.ui.preview;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.view.animation.AccelerateDecelerateInterpolator;

/**
 * @author ZJJ on 2019.10.24
 * */
@SuppressWarnings("unused")
class PreviewFullValueAnimator extends ValueAnimator {

    private boolean isFull;
    private float curDuration;
    private boolean isCancel;

    private FullAnimatorListener listener;

    void start(boolean isFull) {
        if (isRunning()) cancel();
        this.isFull = isFull;
        super.start();
    }

    @Override
    public void cancel() {
        removeAllListeners();
        if (listener != null) listener = null;
        isCancel = true;
        super.cancel();
    }

    PreviewFullValueAnimator(FullAnimatorListener l) {
        this.listener = l;
        setFloatValues(0.0f, 1.0f);
        setInterpolator(new AccelerateDecelerateInterpolator());
        addListener(new AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animation) {
                if (curDuration != 0) curDuration = 0;
            }

            @Override
            public void onAnimationEnd(Animator animation) {
                curDuration = 0;
                if (isCancel) return;
                if (listener != null)
                    listener.onAnimEnd(animation, isFull);
            }

            @Override
            public void onAnimationCancel(Animator animation) {
                curDuration = 0;
            }

            @Override
            public void onAnimationRepeat(Animator animation) {
                curDuration = 0;
            }
        });

        addUpdateListener(new AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator animation) {
                if (isCancel) return;
                if (listener != null) {
                    float duration = (float) animation.getAnimatedValue();
                    float offset = duration - curDuration;
                    listener.onDurationChange(animation, offset, isFull);
                    curDuration = duration;
                }
            }
        });
    }

    public void setAnimatorListener(FullAnimatorListener listener) {
        this.listener = listener;
    }

    public interface FullAnimatorListener {

        void onDurationChange(ValueAnimator animation, float duration, boolean isFull);

        void onAnimEnd(Animator animation, boolean isFull);
    }
}

File: zj_album/src/main/java/com/zj/album/ui/preview/PreviewFullValueAnimator.kt
package com.zj.album.ui.preview

import android.animation.Animator
import android.animation.ValueAnimator
import android.view.animation.AccelerateDecelerateInterpolator

/**
 * @author ZJJ on 2019.10.24
 */
internal class PreviewFullValueAnimator(private var listener: FullAnimatorListener?) : ValueAnimator() {

    private var isFull: Boolean = false
    private var curDuration: Float = 0.toFloat()
    private var isCancel: Boolean = false

    fun start(isFull: Boolean) {
        if (isRunning) cancel()
        this.isFull = isFull
        super.start()
    }

    override fun cancel() {
        removeAllListeners()
        if (listener != null) listener = null
        isCancel = true
        super.cancel()
    }

    init {
        setFloatValues(0.0f, 1.0f)
        interpolator = AccelerateDecelerateInterpolator()
        addListener(object : AnimatorListener {
            override fun onAnimationStart(animation: Animator) {
                if (curDuration != 0f) curDuration = 0f
            }

            override fun onAnimationEnd(animation: Animator) {
                curDuration = 0f
                if (isCancel) return
                listener?.onAnimEnd(animation, isFull)
            }

            override fun onAnimationCancel(animation: Animator) {
                curDuration = 0f
            }

            override fun onAnimationRepeat(animation: Animator) {
                curDuration = 0f
            }
        })

        addUpdateListener(AnimatorUpdateListener { animation ->
            if (isCancel) return@AnimatorUpdateListener
            if (listener != null) {
                val duration = animation.animatedValue as Float
                val offset = duration - curDuration
                listener?.onDurationChange(animation, offset, isFull)
                curDuration = duration
            }
        })
    }

    interface FullAnimatorListener {

        fun onDurationChange(animation: ValueAnimator, duration: Float, isFull: Boolean)

        fun onAnimEnd(animation: Animator, isFull: Boolean)
    }
}

